pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,426 @@
+
+            /* save metadata */
+            packets[metadata_count as usize].start = DEFAULT_HEADROOM;
+            packets[metadata_count as usize].end = DEFAULT_HEADROOM + unsafe { rx_desc_ptr.as_ref().unwrap().len };
+            packets[metadata_count as usize].size = nic.chunk_size;
+            packets[metadata_count as usize].buffer = unsafe { xsk_umem__get_data(nic.buffer, rx_desc_ptr.as_ref().unwrap().addr).cast::<u8>().sub(DEFAULT_HEADROOM as usize) };
+            packets[metadata_count as usize].private = unsafe { (rx_desc_ptr.as_ref().unwrap().addr - DEFAULT_HEADROOM as u64) as *mut c_void };
+
+            // packet_dump(&packets[metadata_count as usize]);
+            metadata_count += 1;
+        }
+
+        unsafe { xsk_ring_cons__release(&mut nic.rx, received); } // notify kernel of using filled slots in RX ring as much as **received
+
+        if metadata_count != received { received = metadata_count; }
+    }
+
+    unsafe {
+        if xsk_ring_prod__needs_wakeup(&mut nic.fq) != 0 {
+            libc::recvfrom(xsk_socket__fd(nic.xsk), std::ptr::null_mut::<libc::c_void>(), 0, libc::MSG_DONTWAIT, std::ptr::null_mut::<libc::sockaddr>(), std::ptr::null_mut::<u32>()) ;
+        }
+    }
+
+    received
+}
+
+pub fn pv_send(nic: &mut PvNic, packets: &mut Vec<PvPacket>, batch_size: u32) -> u32 {
+    /* free packet metadata and UMEM chunks as much as the number of filled slots in cq. */
+    let mut cq_idx: u32 = 0;
+    let filled: u32 = unsafe{ xsk_ring_cons__peek(&mut nic.cq, batch_size, &mut cq_idx) }; // fetch the number of filled slots(the number of packets completely sent) in cq
+
+    if filled > 0 {
+        for _ in 0..filled {
+            unsafe { pv_free_(nic, *xsk_ring_cons__comp_addr(&nic.cq, cq_idx)); } // free UMEM chunks as much as the number of sent packets (same as **filled)
+            cq_idx += 1;
+        }
+        unsafe { xsk_ring_cons__release(&mut nic.cq, filled); } // notify kernel that cq has empty slots with **filled (Dequeue)
+    }
+
+    /* reserve TX ring as much as batch_size before sending packets. */
+    let mut tx_idx: u32 = 0;
+    let reserved: u32 = unsafe { xsk_ring_prod__reserve(&mut nic.tx, batch_size, &mut tx_idx) };
+
+    /* If reservation of slots in TX ring is not sufficient, Don't send them and return 0. */
+    if reserved < batch_size {
+        /* in case that kernel lost interrupt signal in the previous sending packets procedure,
+        repeat to interrupt kernel to send packets which ketnel could have still held.
+        (this procedure is for clearing filled slots in cq, so that cq can be reserved as much as **batch_size in the next execution of pv_send(). */
+        unsafe{
+            if xsk_ring_prod__needs_wakeup(&mut nic.tx) != 0 {
+                libc::sendto(xsk_socket__fd(nic.xsk), std::ptr::null::<libc::c_void>(), 0 as libc::size_t, libc::MSG_DONTWAIT, std::ptr::null::<libc::sockaddr>(), 0 as libc::socklen_t);
```

Comment:
You can break line if one line is too long. And optionally describe what parameter is
```suggestion
                libc::sendto(
                    xsk_socket__fd(nic.xsk),
                    std::ptr::null::<libc::c_void>(),
                    0 as libc::size_t,
                    libc::MSG_DONTWAIT,
                    std::ptr::null::<libc::sockaddr>(),
                    0 as libc::socklen_t);
```